core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java

```diff
@@ -0,0 +1,308 @@
+        String currentFieldName = null;
+        XContentParser.Token token;
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            token = parser.currentToken();
+        } else {
+            token = parser.nextToken();
+        }
+
+        FieldParser<Value> fieldParser = null;
+        if (token != XContentParser.Token.START_OBJECT) {
+            throw new IllegalStateException("[" + name  + "] Expected START_OBJECT but was: " + token);
+        }
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+                fieldParser = getParser(currentFieldName);
+            } else {
+                if (currentFieldName == null) {
+                    throw new IllegalStateException("[" + name  + "] no field found");
+                }
+                assert fieldParser != null;
+                fieldParser.assertSupports(name, token, currentFieldName, parser.getParseFieldMatcher());
+                parseNextToken(parser, fieldParser, currentFieldName, value, context);
+                fieldParser = null;
+            }
+        }
+        return value;
+    }
+
+    private void parseArray(XContentParser parser, FieldParser<Value> fieldParser, String currentFieldName, Value value, Context context) throws IOException {
+        assert parser.currentToken() == XContentParser.Token.START_ARRAY : "Token was: " + parser.currentToken();
+        parseValue(parser, fieldParser, currentFieldName, value, context);
+    }
+
+    private void parseValue(XContentParser parser, FieldParser<Value> fieldParser, String currentFieldName, Value value, Context context) throws IOException {
+        try {
+            fieldParser.parser.parse(parser, value, context);
+        } catch (Throwable ex) {
+            throw new ParsingException(new Index("_na_"), parser, "[" + name  + "] failed to parse field [" + currentFieldName + "]", ex);
+        }
+    }
+
+    private void parseNextToken(XContentParser parser, FieldParser<Value> fieldParser, String currentFieldName, Value value, Context context) throws IOException {
+        final XContentParser.Token token = parser.currentToken();
+        switch (token) {
+            case START_OBJECT:
+                parseValue(parser, fieldParser, currentFieldName, value, context);
+                break;
+            case START_ARRAY:
+                parseArray(parser, fieldParser, currentFieldName, value, context);
+                break;
+            case END_OBJECT:
+            case END_ARRAY:
+            case FIELD_NAME:
+                throw new IllegalStateException("[" + name  + "]" + token + " is unexpected");
+            case VALUE_STRING:
+            case VALUE_NUMBER:
+            case VALUE_BOOLEAN:
+            case VALUE_EMBEDDED_OBJECT:
+            case VALUE_NULL:
```

Comment:
add a default case for unknown? so if the enum ever changes, we aren't caught with a weird error.